File: transports/ocaml/lib/events.ml
(* See events.mli for the contract. Ported from transports/go/events.go. *)

open Conventions

type profile = Verbose | Compact

let profile_name = function Verbose -> "verbose" | Compact -> "compact"

let parse_profile = function
  | "verbose" -> Some Verbose
  | "compact" -> Some Compact
  | _ -> None

type event = {
  service : string option;
  service_ord : int64 option;
  event : string option;
  op_ord : int64 option;
  id : int64 option;
  payload : bytes;
}

let new_verbose_event service event payload =
  {
    service;
    service_ord = None;
    event = Some event;
    op_ord = None;
    id = None;
    payload;
  }

let new_compact_event service_ord op_ord payload =
  {
    service = None;
    service_ord = Some service_ord;
    event = None;
    op_ord = Some op_ord;
    id = None;
    payload;
  }

let event_with_id e id = { e with id = Some id }

let encode_verbose (e : event) =
  match e.event with
  | None -> Error (malformed "verbose event missing 'event' name")
  | Some ev ->
      let entries =
        [
          (Cbor.Text "event", Cbor.Text ev);
          (Cbor.Text "payload", tag24 e.payload);
        ]
      in
      let entries =
        match e.service with
        | Some s -> entries @ [ (Cbor.Text "service", Cbor.Text s) ]
        | None -> entries
      in
      let entries =
        match e.id with
        | Some id -> entries @ [ (Cbor.Text "id", Cbor.Uint id) ]
        | None -> entries
      in
      Ok (Cbor.encode (Cbor.canon_map entries))

let encode_compact (e : event) =
  match (e.service_ord, e.op_ord) with
  | None, _ -> Error (malformed "compact event missing service ordinal")
  | _, None -> Error (malformed "compact event missing op ordinal")
  | Some so, Some oo ->
      let head = [ Cbor.Uint so; Cbor.Uint oo ] in
      let with_id =
        match e.id with Some id -> head @ [ Cbor.Uint id ] | None -> head
      in
      Ok (Cbor.encode (Cbor.Array (with_id @ [ tag24 e.payload ])))

let encode_event e = function
  | Verbose -> encode_verbose e
  | Compact -> encode_compact e

let decode_verbose b =
  let* v = decode_envelope b in
  match map_get v "payload" with
  | None -> Error (malformed "missing 'payload'")
  | Some p ->
      let* payload = untag24 p in
      let* ev = get_text v "event" in
      Ok
        {
          service = get_text_opt v "service";
          service_ord = None;
          event = Some ev;
          op_ord = None;
          id = get_uint_opt v "id";
          payload;
        }

let ord_of v =
  match as_u64 v with
  | Some n -> Ok n
  | None -> Error (malformed "ordinal is not an integer")

let decode_compact b =
  let* v = decode_envelope b in
  match v with
  | Cbor.Array arr -> (
      match arr with
      | [ so; oo; pl ] ->
          let* service_ord = ord_of so in
          let* op_ord = ord_of oo in
          let* payload = untag24 pl in
          Ok
            {
              service = None;
              service_ord = Some service_ord;
              event = None;
              op_ord = Some op_ord;
              id = None;
              payload;
            }
      | [ so; oo; idv; pl ] ->
          let* service_ord = ord_of so in
          let* op_ord = ord_of oo in
          let* id = ord_of idv in
          let* payload = untag24 pl in
          Ok
            {
              service = None;
              service_ord = Some service_ord;
              event = None;
              op_ord = Some op_ord;
              id = Some id;
              payload;
            }
      | _ ->
          Error
            (malformed "compact event array has %d elements, expected 3 or 4"
               (List.length arr)))
  | _ -> Error (malformed "compact event is not an array")

let decode_event b = function
  | Verbose -> decode_verbose b
  | Compact -> decode_compact b

let control_hello = 0L
let control_hello_ack = 1L
let control_ping = 2L
let control_pong = 3L
let control_close = 4L
let control_error = 5L
let hello_name = "$hello"
let hello_ack_name = "$hello-ack"
let ping_name = "$ping"
let pong_name = "$pong"
let close_name = "$close"
let error_name = "$error"

type hello = {
  versions : int64 list;
  profiles : string list;
  hello_service : string option;
  hello_auth : string option;
}

let encode_hello (h : hello) =
  let versions = List.map (fun v -> Cbor.Uint v) h.versions in
  let profiles = List.map (fun p -> Cbor.Text p) h.profiles in
  let entries =
    [
      (Cbor.Text "versions", Cbor.Array versions);
      (Cbor.Text "profiles", Cbor.Array profiles);
    ]
  in
  let entries =
    match h.hello_service with
    | Some s -> entries @ [ (Cbor.Text "service", Cbor.Text s) ]
    | None -> entries
  in
  let entries =
    match h.hello_auth with
    | Some a -> entries @ [ (Cbor.Text "auth", Cbor.Text a) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_hello b =
  let* v = decode_envelope b in
  let* versions =
    match map_get v "versions" with
    | Some (Cbor.Array xs) -> Ok (List.filter_map as_u64 xs)
    | _ -> Error (malformed "hello missing 'versions'")
  in
  let* profiles =
    match map_get v "profiles" with
    | Some (Cbor.Array xs) ->
        Ok (List.filter_map (function Cbor.Text t -> Some t | _ -> None) xs)
    | _ -> Error (malformed "hello missing 'profiles'")
  in
  Ok
    {
      versions;
      profiles;
      hello_service = get_text_opt v "service";
      hello_auth = get_text_opt v "auth";
    }

let negotiate (h : hello) supported =
  if not (List.exists (fun v -> Int64.equal v version) h.versions) then None
  else
    let rec find = function
      | [] -> None
      | offered :: rest -> (
          match parse_profile offered with
          | Some p when List.mem p supported -> Some (version, p)
          | _ -> find rest)
    in
    find h.profiles

type hello_ack = { v : int64; ack_profile : string; session : string option }

let encode_hello_ack (a : hello_ack) =
  let entries =
    [
      (Cbor.Text "v", Cbor.Uint a.v);
      (Cbor.Text "profile", Cbor.Text a.ack_profile);
    ]
  in
  let entries =
    match a.session with
    | Some s -> entries @ [ (Cbor.Text "session", Cbor.Text s) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_hello_ack b =
  let* v = decode_envelope b in
  let* ver = get_uint v "v" in
  let* prof = get_text v "profile" in
  Ok { v = ver; ack_profile = prof; session = get_text_opt v "session" }

type heartbeat = { nonce : int64; at : int64 option }

let encode_heartbeat (h : heartbeat) =
  let entries = [ (Cbor.Text "nonce", Cbor.Uint h.nonce) ] in
  let entries =
    match h.at with
    | Some a -> entries @ [ (Cbor.Text "at", Cbor.Uint a) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_heartbeat b =
  let* v = decode_envelope b in
  let* nonce = get_uint v "nonce" in
  Ok { nonce; at = get_uint_opt v "at" }

type close = { status : Conventions.status; reason : string option }

let encode_close (c : close) =
  let entries = [ (Cbor.Text "status", Cbor.Int (status_code c.status)) ] in
  let entries =
    match c.reason with
    | Some r -> entries @ [ (Cbor.Text "reason", Cbor.Text r) ]
    | None -> entries
  in
  Cbor.encode (Cbor.canon_map entries)

let decode_close b =
  let* v = decode_envelope b in
  let* status = get_int v "status" in
  Ok { status = status_from_code status; reason = get_text_opt v "reason" }

File: transports/dart/lib/src/events.dart
// CSIL-Events transport — typed bidirectional event streams — see
// `csil-events-transport.md`. Verbose (text-keyed) and compact (positional)
// profiles, plus the control plane (service ordinal 0) lifecycle payloads.

import 'dart:typed_data';

import 'cbor.dart';
import 'cbor.dart' as cbor show encode, decode;
import 'conventions.dart';

/// Which wire profile a connection uses for its lifetime, fixed by `hello`. A
/// genuinely closed two-member set, so a Dart `enum` (unlike the transport
/// `Status` registry) is the right model.
enum Profile { verbose, compact }

Profile? parseProfile(String s) => switch (s) {
      'verbose' => Profile.verbose,
      'compact' => Profile.compact,
      _ => null,
    };

/// Control-plane operation ordinals (under service ordinal 0) and their `$`-sigil
/// verbose names.
abstract final class Control {
  static const int hello = 0;
  static const int helloAck = 1;
  static const int ping = 2;
  static const int pong = 3;
  static const int close = 4;
  static const int error = 5;

  static const String helloName = r'$hello';
  static const String helloAckName = r'$hello-ack';
  static const String pingName = r'$ping';
  static const String pongName = r'$pong';
  static const String closeName = r'$close';
  static const String errorName = r'$error';
}

/// One typed event flowing in either direction. Identified by service+operation
/// (verbose) or by their ordinals (compact); carries an optional correlation `id`
/// when it is a request expecting a reply, or that reply.
final class Event {
  String?
      service; // CSIL service name (verbose); omitted on a single-service connection
  int?
      serviceOrd; // service ordinal (compact); always present in compact frames
  String? event; // CSIL operation name (verbose)
  int? opOrd; // operation ordinal (compact)
  int? id;
  Uint8List payload;

  Event._(this.payload);

  /// A verbose event by name. `service` is omitted on a single-service connection.
  factory Event.verbose(String? service, String event, Uint8List payload) {
    final e = Event._(payload);
    e.service = service;
    e.event = event;
    return e;
  }

  /// A compact event by ordinals.
  factory Event.compact(int serviceOrd, int opOrd, Uint8List payload) {
    final e = Event._(payload);
    e.serviceOrd = serviceOrd;
    e.opOrd = opOrd;
    return e;
  }

  Event withId(int id) {
    this.id = id;
    return this;
  }

  Uint8List encode(Profile profile) =>
      profile == Profile.verbose ? _encodeVerbose() : _encodeCompact();

  Uint8List _encodeVerbose() {
    final name = event;
    if (name == null) {
      throw const MalformedException("verbose event missing 'event' name");
    }
    final entries = <(String, Cbor)>[
      ('event', CborText(name)),
      ('payload', tag24(payload)),
    ];
    if (service != null) entries.add(('service', CborText(service!)));
    if (id != null) entries.add(('id', CborInt(id!)));
    return cbor.encode(canonMap(entries));
  }

  Uint8List _encodeCompact() {
    final sOrd = serviceOrd;
    final oOrd = opOrd;
    if (sOrd == null) {
      throw const MalformedException('compact event missing service ordinal');
    }
    if (oOrd == null) {
      throw const MalformedException('compact event missing op ordinal');
    }
    final arr = <Cbor>[CborInt(sOrd), CborInt(oOrd)];
    if (id != null) arr.add(CborInt(id!));
    arr.add(tag24(payload));
    return cbor.encode(CborArray(arr));
  }

  static Event decode(Uint8List bytes, Profile profile) =>
      profile == Profile.verbose
          ? _decodeVerbose(bytes)
          : _decodeCompact(bytes);

  static Event _decodeVerbose(Uint8List bytes) {
    final v = cbor.decode(bytes);
    final payloadField = mapGet(v, 'payload');
    if (payloadField == null) {
      throw const MalformedException("missing 'payload'");
    }
    final e = Event.verbose(
        getTextOpt(v, 'service'), getText(v, 'event'), untag24(payloadField));
    e.id = getUintOpt(v, 'id');
    return e;
  }

  static Event _decodeCompact(Uint8List bytes) {
    final v = cbor.decode(bytes);
    if (v is! CborArray) {
      throw const MalformedException('compact event is not an array');
    }
    final arr = v.value;
    // 3 elements => [service_ord, op_ord, payload]; 4 => with correlation id.
    Cbor payloadVal;
    Cbor? idVal;
    if (arr.length == 3) {
      payloadVal = arr[2];
    } else if (arr.length == 4) {
      idVal = arr[2];
      payloadVal = arr[3];
    } else {
      throw MalformedException(
          'compact event array has ${arr.length} elements, expected 3 or 4');
    }
    final e =
        Event.compact(_asUint(arr[0]), _asUint(arr[1]), untag24(payloadVal));
    if (idVal != null) e.id = _asUint(idVal);
    return e;
  }

  static int _asUint(Cbor val) {
    if (val is CborInt && val.value >= 0) return val.value;
    throw const MalformedException('ordinal is not a non-negative integer');
  }

  @override
  bool operator ==(Object other) =>
      other is Event &&
      other.service == service &&
      other.serviceOrd == serviceOrd &&
      other.event == event &&
      other.opOrd == opOrd &&
      other.id == id &&
      bytesEqual(other.payload, payload);

  @override
  int get hashCode =>
      Object.hash(service, serviceOrd, event, opOrd, id, payload.length);
}

/// The `$hello` payload offered by the connection initiator.
final class Hello {
  final List<int> versions;
  final List<String> profiles;
  String? service;
  String? auth;

  Hello(this.versions, this.profiles, {this.service, this.auth});

  Uint8List encode() {
    final entries = <(String, Cbor)>[
      ('versions', CborArray(versions.map<Cbor>((n) => CborInt(n)).toList())),
      ('profiles', CborArray(profiles.map<Cbor>((p) => CborText(p)).toList())),
    ];
    if (service != null) entries.add(('service', CborText(service!)));
    if (auth != null) entries.add(('auth', CborText(auth!)));
    return cbor.encode(canonMap(entries));
  }

  static Hello decode(Uint8List bytes) {
    final v = cbor.decode(bytes);
    final versionsField = mapGet(v, 'versions');
    if (versionsField is! CborArray) {
      throw const MalformedException("hello missing 'versions'");
    }
    final versions = versionsField.value
        .whereType<CborInt>()
        .map((x) => x.value)
        .toList(growable: false);
    final profilesField = mapGet(v, 'profiles');
    if (profilesField is! CborArray) {
      throw const MalformedException("hello missing 'profiles'");
    }
    final profiles = profilesField.value
        .whereType<CborText>()
        .map((x) => x.value)
        .toList(growable: false);
    return Hello(versions, profiles,
        service: getTextOpt(v, 'service'), auth: getTextOpt(v, 'auth'));
  }

  /// Select a profile from this hello's offers, honoring the peer's preference
  /// order and what the server supports. Returns `(version, profile)`, or `null`
  /// if nothing is mutually supported.
  (int, Profile)? negotiate(List<Profile> supported) {
    if (!versions.contains(version)) return null;
    for (final offered in profiles) {
      final p = parseProfile(offered);
      if (p != null && supported.contains(p)) return (version, p);
    }
    return null;
  }
}

/// The `$hello-ack` payload returned by the peer.
final class HelloAck {
  final int v;
  final String profile;
  String? session;

  HelloAck(this.v, this.profile, {this.session});

  Uint8List encode() {
    final entries = <(String, Cbor)>[
      ('v', CborInt(v)),
      ('profile', CborText(profile)),
    ];
    if (session != null) entries.add(('session', CborText(session!)));
    return cbor.encode(canonMap(entries));
  }

  static HelloAck decode(Uint8List bytes) {
    final value = cbor.decode(bytes);
    return HelloAck(getUint(value, 'v'), getText(value, 'profile'),
        session: getTextOpt(value, 'session'));
  }
}

/// A `$ping`/`$pong` heartbeat payload.
final class Heartbeat {
  final int nonce;
  int? at;

  Heartbeat(this.nonce, {this.at});

  Uint8List encode() {
    final entries = <(String, Cbor)>[('nonce', CborInt(nonce))];
    if (at != null) entries.add(('at', CborInt(at!)));
    return cbor.encode(canonMap(entries));
  }

  static Heartbeat decode(Uint8List bytes) {
    final v = cbor.decode(bytes);
    return Heartbeat(getUint(v, 'nonce'), at: getUintOpt(v, 'at'));
  }
}

/// A `$close` payload. `status` is a transport-registry code.
final class Close {
  final Status status;
  String? reason;

  Close(this.status, {this.reason});

  Uint8List encode() {
    final entries = <(String, Cbor)>[('status', CborInt(status.code))];
    if (reason != null) entries.add(('reason', CborText(reason!)));
    return cbor.encode(canonMap(entries));
  }

  static Close decode(Uint8List bytes) {
    final v = cbor.decode(bytes);
    return Close(Status(getInt(v, 'status')), reason: getTextOpt(v, 'reason'));
  }
}
